parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/WorkManager.java

```diff
@@ -299,19 +292,42 @@ public boolean isDirty() {
         return pm.isDirty();
     }
 
-    public void addMetricsTo(PCMetrics.PCMetricsBuilder<?, ? extends PCMetrics.PCMetricsBuilder<?, ?>> metrics) {
-        metrics.partitionMetrics(getPm().getMetrics());
-        metrics.shardMetrics(getSm().getMetrics());
-        metrics.workManagerMetrics(getMetrics());
+    private void initMetrics() {
+        waitingRecordsNumberGauge = PCMetrics.getInstance().gaugeFromMetricDef(PCMetricsDef.WAITING_RECORDS,
+                this, WorkManager::getNumberOfWorkQueuedInShardsAwaitingSelection);
+        inflightRecordsNumberGauge = PCMetrics.getInstance().gaugeFromMetricDef(PCMetricsDef.INFLIGHT_RECORDS,
+                this, WorkManager::getNumberRecordsOutForProcessing);
     }
 
-    public void enrichWithIncompletes(Map<TopicPartition, PCMetrics.PCPartitionMetrics> metrics) {
-        getPm().enrichWithIncompletes(metrics);
+    private void initCounterMetrics(Collection<TopicPartition> partitions) {
+        partitions.forEach(topicPartition -> {
+            if (!succeededRecordsCounters.containsKey(topicPartition)) {
+                succeededRecordsCounters.put(topicPartition, PCMetrics.getInstance().getCounterFromMetricDef(PCMetricsDef.PROCESSED_RECORDS, getWorkManagerCounterTags(topicPartition)));
+            }
+            if (!failedRecordsCounters.containsKey(topicPartition)) {
+                failedRecordsCounters.put(topicPartition, PCMetrics.getInstance().getCounterFromMetricDef(PCMetricsDef.FAILED_RECORDS, getWorkManagerCounterTags(topicPartition)));
+            }
+        });
+    }
+
+    private void safelyIncrementCounter(Map<TopicPartition, Counter> counterMap, TopicPartition topicPartition) {
+        Optional.ofNullable(counterMap.get(topicPartition)).ifPresent(Counter::increment);
     }
 
-    private PCMetrics.WorkManagerMetrics getMetrics() {
-        return PCMetrics.WorkManagerMetrics.builder().
-        waitingRecords(getNumberOfWorkQueuedInShardsAwaitingSelection())
-                .inflightRecords(getNumberRecordsOutForProcessing()).build();
+    private Tag[] getWorkManagerCounterTags(TopicPartition topicPartition) {
+        return new Tag[]{Tag.of("topic", topicPartition.topic()), Tag.of("partition", String.valueOf(topicPartition.partition()))};
+    }
+
+    private void deregisterCounterMetrics(Collection<TopicPartition> partitions) {
```

Comment:
Should we make this more generic like `deregisterTopicPartitionMetrics`? There could be other types in the future. As an aside -- do we need to deregister those gauges here too?

Reply:
Yep will rename.
Gauges dont need to be deregistered - only topic / partition specific metrics need created / deregistered - as topic / partition is assigned / revoked. 
Gauges are tracking overall count of inflight / waiting messages on WorkManager level so they are fine.